### PR TITLE
Lostromos should have an opt in mode

### PR DESCRIPTION
**Pull request by wpdrush21**
Signed-off-by: Derek Rushing <derek.rushing21@gmail.com>

# What Are We Doing Here

Adding the ability to 'opt-in' for Lostromos use by CRs. If given opt-in criteria, custom resources should have an annotation with the given opt-in key in order to have the watcher pass on information to the controllers.

## How to Verify

1. Check out this PR and run `make build`.
2. Start minikube and run `kubectl apply -f test/data/crd.yml`.
3. Run `kubectl apply -f test/data/cr_things.yml`
4. Finally run `./lostromos start --config test/data/config.yaml --nop --crd-optIn optIn` and see that the resources aren't printed.
5. Run `kubectl delete -f test/data/cr_things.yml`
6. Add an annotation to Thing1 with key/values of `optIn: "true"`
7. Re-run step 3 and see that Lostromos now prints out Thing1

### CODE REVIEW

File: crwatcher/watcher_test.go [hunk 1]
```diff
@@ -99,6 +101,44 @@ func TestSetupHandlerAddFunc(t *testing.T) {
...
+// Test to ensure that if we are given opt in criteria we only call ResourceAdded for a resource with the specified opt
+// in criteria.
+func TestSetupHandlerAddFuncOptIn(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	mockRC := NewMockResourceController(mockCtrl)
+	cw := &CRWatcher{
+		Config: &Config{
+			OptInKey: "optIn",
```
Comment (nicolerenee) [inline crwatcher/watcher_test.go]: Since people will use tests as examples of it. Can we change `optIn` to `com.wpengine.lostromos.optIn`

Comment (nicolerenee) [reply crwatcher/watcher_test.go]: And make this change everywhere. Annotations are suppose to follow a dns naming pattern.

File: crwatcher/watcher.go [hunk 2]
```diff
@@ -114,15 +115,24 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
...
 			r := obj.(*unstructured.Unstructured)
+			if cw.Config.OptInKey != "" && !cw.hasOptInKey(r) {
+				return
+			}
 			con.ResourceDeleted(r)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			oldR := oldObj.(*unstructured.Unstructured)
 			newR := newObj.(*unstructured.Unstructured)
+			if cw.Config.OptInKey != "" && (!cw.hasOptInKey(oldR) && !cw.hasOptInKey(newR)) {
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: We should only check if the `newR` has the optIn key here, I think. 

An interesting scenario though is what happens where the old has the optIn key and the new record doesn't. Should that trigger a delete?

File: cmd/start.go [hunk 3]
```diff
@@ -54,6 +54,7 @@ func init() {
 	startCmd.Flags().String("crd-group", "", "the group of the CRD you want monitored (ex: stable.wpengine.io)")
 	startCmd.Flags().String("crd-version", "v1", "the version of the CRD you want monitored")
 	startCmd.Flags().String("crd-namespace", metav1.NamespaceNone, "(optional) the namespace of the CRD you want monitored, only needed for namespaced CRDs (ex: default)")
+	startCmd.Flags().String("crd-optIn", "", "(optional) Annotation key to specify that the custom resource has opted in to watching by Lostromos")
```
Comment (nicolerenee) [inline cmd/start.go]: To document our conversation around this, `crd-filter` appears to be a better name for this.

File: crwatcher/watcher.go (hunk 2 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: We decided that optIn on the old, but not on the new should perform a Delete. All other cases should act as expected

No && No => No-Op
OptIn && OptIn => Update
No && OptIn => Update

This will be documented in an events doc.

File: crwatcher/watcher.go [hunk 4]
```diff
@@ -114,20 +116,47 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
...
+// is valid here, but we've decided on doing a delete.
+func (cw *CRWatcher) filteredUpdate(con ResourceController, oldR *unstructured.Unstructured, newR *unstructured.Unstructured) {
+	if !cw.hasFilter(oldR) && !cw.hasFilter(newR) {
+		return
+	} else if cw.hasFilter(oldR) && !cw.hasFilter(newR) {
+		con.ResourceDeleted(oldR)
+	} else {
+		// This acts as a type of create, so we need to add a new managed release.
+		if !cw.hasFilter(oldR) && cw.hasFilter(newR) {
+			metrics.ManagedReleases.Inc()
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: We need to not have metrics in this class. This is a packages that can be imported into other programs and it should not require internal things to happen. In this case we are doing a create so you should just call the ResourceCreated function. The create function already has to handle being called for a resource that already exists.

File: crwatcher/watcher.go [hunk 5]
```diff
@@ -152,6 +181,18 @@ func (cw *CRWatcher) setupController() {
 	)
 }
 
+// hasFilter checks to see if we are using an opt in filter (if not, then return true), and if so returns whether we
+// have an annotation matching the given filter.
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: This documentation is incorrect. We don't check to see if we are using a filter in here and return true. Should we be? Also based on my other commends can we make this `clearsFilter` and it also checks if there is even a filter in play.

File: crwatcher/watcher.go [hunk 6]
```diff
@@ -114,20 +116,47 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
 	cw.handler = cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			r := obj.(*unstructured.Unstructured)
+			if cw.Config.Filter != "" && !cw.hasFilter(r) {
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: Could we move the check for the filter into the method we call. Then update it to work like the documentation says it will work? Maybe make this somehting like 
``` 
if cw.clearsFilter(r) {
    ..do stuff..
}
```
Then clearsFilter returns true if the resource is not filtered out. I think this helps a lot with readability not having to have `!hasFilter`

File: crwatcher/watcher.go [hunk 7]
```diff
@@ -114,20 +116,47 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
...
+				return
+			}
 			con.ResourceDeleted(r)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
 			oldR := oldObj.(*unstructured.Unstructured)
 			newR := newObj.(*unstructured.Unstructured)
-			con.ResourceUpdated(oldR, newR)
+			if cw.Config.Filter != "" {
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: can we remove all the logic here and have this only call our update and that triggers the external c/u/d

File: crwatcher/watcher.go (hunk 4 above)
Comment (nicolerenee) [reply crwatcher/watcher.go]: Actually going through this again, this should just call update. We should add logic on the update side so that if it creates a resource during update it will increment the managed releases counter. This will be easy to do on the helm controller, and harder to do on the template controller. But I think that is acceptable.

File: crwatcher/watcher.go [hunk 8]
```diff
@@ -114,20 +116,47 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
...
+				con.ResourceUpdated(oldR, newR)
+			}
 		},
 	}
 }
 
+// filteredUpdate performs an updated based on the filter status of the given resource. In the case that we have a
+// filter match for the old resource, but not for the new, we should delete the old resource. Either a no-op or delete
+// is valid here, but we've decided on doing a delete.
+func (cw *CRWatcher) filteredUpdate(con ResourceController, oldR *unstructured.Unstructured, newR *unstructured.Unstructured) {
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: This method is kind of hard to follow and I know what it should be doing.  Can we make it look something like
```
if cw.clearsFilter(newR) {
    con.ResourceUpdated(oldR, newR)
  } else if cw.clearsFilter(oldR) {
    // The resource is filtered out and needs to be removed as it previously existed
    con.ResourceDeleted(oldR)
  }
```
This might not be better for readability so I'm open to suggestions.

File: crwatcher/watcher.go (hunk 6 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: Done

File: crwatcher/watcher.go (hunk 7 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: Done

File: crwatcher/watcher.go (hunk 8 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: I agree, I actually hadn't seen this until I re-organized and came up with the same exact thing. To me that suggests it's the proper approach :)

File: crwatcher/watcher.go (hunk 5 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: Changed to clearsFilter, and the docs are changed accordingly.

File: crwatcher/watcher_test.go [hunk 9]
```diff
@@ -99,12 +101,52 @@ func TestSetupHandlerAddFunc(t *testing.T) {
 	cw.handler.OnAdd(r)
 }
 
+// Test to ensure that if we are given opt in criteria we only call ResourceAdded for a resource with the specified opt
+// in criteria.
+func TestSetupHandlerAddFuncOptIn(t *testing.T) {
```
Comment (nicolerenee) [inline crwatcher/watcher_test.go]: This should use the name filter not OptIn for the test name and the documentation above updated as well.

File: crwatcher/watcher_test.go [hunk 10]
```diff
@@ -119,12 +161,52 @@ func TestSetupHandlerDeleteFunc(t *testing.T) {
 	cw.handler.OnDelete(r)
 }
 
+// Test to ensure that if we are given opt in criteria we only call ResourceDeleted for a resource with the specified
+// opt in criteria.
+func TestSetupHandlerDeleteFuncOptIn(t *testing.T) {
```
Comment (nicolerenee) [inline crwatcher/watcher_test.go]: Same thing here.

File: crwatcher/watcher_test.go [hunk 11]
```diff
@@ -141,9 +223,68 @@ func TestSetupHandlerUpdateFunc(t *testing.T) {
...
 
-	mockRC.EXPECT().ResourceUpdated(r1, r2)
+	mockRC.EXPECT().ResourceUpdated(r1, r2, false)
+
+	cw.handler.OnUpdate(r1, r2)
+}
+
+// Test to ensure that if we are given opt in criteria we only call ResourceUpdated for resources with the specified
+// opt in criteria.
+func TestSetupHandlerUpdateFuncOptIn(t *testing.T) {
```
Comment (nicolerenee) [inline crwatcher/watcher_test.go]: and again here.

File: docs/events.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
+# Understanding Lostromos Events
+
+This document is meant to describe the events that will occur when you use Lostromos with different settings.
+
+## Updates with a Filter
+
+When performing updates and using the `crd.filter` option, we have defined the behavior:
+
+* Old Resource: No Filter, New Resource: No Filter - No-op
```
Comment (nicolerenee) [inline docs/events.md]: Maybe instead of "No Filter" and "Filter" we can use "Filter annotation exists" and "Filter annotation doesn't exist". Makes it a little more clear.

File: docs/workingWithLostromos.md [hunk 13]
```diff
@@ -29,6 +29,7 @@ To start working with Lostrómos, you should begin by playing around with some b
   * `group` (Required) The group of the CRD you want monitored (ex: stable.wpengine.io)
   * `version` (Required) The version of the CRD you want monitored
   * `namespace` The namespace of the CRD you want monitored
+  * `filter` Filter to specify if Lostromos will act on a resource create/update/delete
```
Comment (nicolerenee) [inline docs/workingWithLostromos.md]: Probably should include a link to the events doc here for help understanding how this works.

File: helmctlr/helm.go [hunk 14]
```diff
@@ -100,6 +100,9 @@ func (c Controller) ResourceUpdated(oldR, newR *unstructured.Unstructured) {
 		return
 	}
 	metrics.UpdatedReleases.Inc()
+	if newResourceAfterUpdate {
```
Comment (nicolerenee) [inline helmctlr/helm.go]: We should add tests for this.

File: crwatcher/watcher.go [hunk 15]
```diff
@@ -65,7 +66,7 @@ type CRWatcher struct {
 //      don't notice the deletion until the subsequent re-list.
 type ResourceController interface {
 	ResourceAdded(resource *unstructured.Unstructured)
-	ResourceUpdated(oldResource, newResource *unstructured.Unstructured)
+	ResourceUpdated(oldResource, newResource *unstructured.Unstructured, newResourceAfterUpdate bool)
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: I don't love that we are changing this interface. I kind of lean toward making the controllers smarter rather than doing this, or instead of calling Updated we call Added.

File: test/data/cr_things_filter_update.yml [hunk 16]
```diff
@@ -0,0 +1,37 @@
+---
+
+apiVersion: stable.nicolerenee.io/v1
+kind: Character
+metadata:
+  name: thing1
+spec:
+  Name: Thing 1
+  From: Cat in the Hat
+  By: Dr. Suess
```
Comment (nicolerenee) [inline test/data/cr_things_filter_update.yml]: This is totally my fault...but this is spelled wrong. It's Dr. Seuss

Comment (wpdrush21) [reply test/data/cr_things_filter_update.yml]: Fixed

File: helmctlr/helm.go (hunk 14 above)
Comment (wpdrush21) [reply helmctlr/helm.go]: Fixed from changing to do a ResourceAdded instead of an update.

File: crwatcher/watcher.go (hunk 15 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: Changed to calling Added.

File: crwatcher/watcher_test.go (hunk 9 above)
Comment (wpdrush21) [reply crwatcher/watcher_test.go]: Done

File: crwatcher/watcher_test.go (hunk 10 above)
Comment (wpdrush21) [reply crwatcher/watcher_test.go]: Done

File: crwatcher/watcher_test.go (hunk 11 above)
Comment (wpdrush21) [reply crwatcher/watcher_test.go]: Done

File: docs/events.md (hunk 12 above)
Comment (wpdrush21) [reply docs/events.md]: Changed, and moved this to a table instead.

File: docs/workingWithLostromos.md (hunk 13 above)
Comment (wpdrush21) [reply docs/workingWithLostromos.md]: Done

File: crwatcher/watcher.go [hunk 17]
```diff
@@ -114,20 +115,44 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
...
 	}
 }
 
+// update sends an appropriate notification to the controller based on filtering outcomes of the old and new state of a
+// resource.
+//
+// If no filter is configured or both states of the resource pass filtering, send an update to the controller.
+// If the new state passes filtering and the old state does not, send an add notification to the controller.
+// If the old state passes filtering and the new state does not, send a delete notification to the controller.
+// If neither state passes filtering, ignore.
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: Really good docs here!

File: crwatcher/watcher.go [hunk 18]
```diff
@@ -114,20 +115,44 @@ func (cw *CRWatcher) setupHandler(con ResourceController) {
...
+//
+func (cw *CRWatcher) update(con ResourceController, oldR *unstructured.Unstructured, newR *unstructured.Unstructured) {
+	if cw.noFilterOrPassesFiltering(newR) {
+		if cw.noFilterOrPassesFiltering(oldR) {
+			con.ResourceUpdated(oldR, newR)
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: We can make this `return con.ResourceUpdated(oldR, newR)` and then we can get rid of the else. 

Go lint normally catches this, but since we don't have an explicit return it doesn't. Normally you want code like
```
func get(i int) int {
    f := i * 2
    if f == 6 {
        return f
    }
    return f * 3
}
```

File: crwatcher/watcher.go [hunk 19]
```diff
@@ -152,6 +177,22 @@ func (cw *CRWatcher) setupController() {
 	)
 }
 
+// noFilterOrPassesFiltering checks to see if we are using an opt in filter (if not, then return true), and if so returns whether we
```
Comment (nicolerenee) [inline crwatcher/watcher.go]: This name is a lot longer than it was in the previous version, and starts to hit my internal thought of it's against go's mentality of short names. I can be convinced to leave it with any bit of objection against the shorter name.

File: docs/events.md [hunk 20]
```diff
@@ -0,0 +1,16 @@
+# Understanding Lostromos Events
+
+This document is meant to describe the events that will occur when you use Lostromos with different settings.
+
+## Updates with a Filter
+
+When performing updates and using a non empty filter via the `crd.filter` option, we have defined the behavior:
+
+| Old Resource | New Resource | Action Taken |
```
Comment (nicolerenee) [inline docs/events.md]: This is way better as a table. Good call!

File: printctlr/controller.go [hunk 21]
```diff
@@ -39,9 +39,9 @@ func (c Controller) ResourceAdded(r *unstructured.Unstructured) {
 // ResourceUpdated receives both an the old version and current version of a
 // custom resource and will print out the the custom resource was changed
 func (c Controller) ResourceUpdated(oldR, newR *unstructured.Unstructured) {
+	metrics.TotalEvents.Inc()
```
Comment (nicolerenee) [inline printctlr/controller.go]: Any particular reason for moving this to the top? Just wanting to understand.

Comment (wpdrush21) [reply printctlr/controller.go]: Fixed

File: crwatcher/watcher.go (hunk 18 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: Changed

File: crwatcher/watcher.go (hunk 19 above)
Comment (wpdrush21) [reply crwatcher/watcher.go]: Shortened

File: test/data/cr_things_filter.yml [hunk 22]
```diff
@@ -0,0 +1,36 @@
+---
+
+apiVersion: stable.nicolerenee.io/v1
+kind: Character
+metadata:
+  name: thing1
+  annotations:
+    optIn: Opting In
```
Comment (nicolerenee) [inline test/data/cr_things_filter.yml]: Missed this before. Can we update this example with a full DNS example so that it is valid. Maybe here use `io.nicolerenee.lostromosApplied` or something not under the `com.wpengine.lostromos` so that people will see it can be any value. But it should be DNS

File: test/data/cr_things_filter.yml [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+  By: Dr. Seuss
+
+---
+
+apiVersion: stable.nicolerenee.io/v1
+kind: Character
+metadata:
+  name: thing3
+  annotations:
+    optIn: Opting In Again
```
Comment (nicolerenee) [inline test/data/cr_things_filter.yml]: and here

File: test/data/cr_things_filter_update.yml [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+  By: Dr. Seuss
+
+---
+
+apiVersion: stable.nicolerenee.io/v1
+kind: Character
+metadata:
+  name: thing2
+  annotations:
+    optIn: Opting In
```
Comment (nicolerenee) [inline test/data/cr_things_filter_update.yml]: and here

File: test/data/cr_things_filter_update.yml [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+  By: Dr. Seuss
+
+---
+
+apiVersion: stable.nicolerenee.io/v1
+kind: Character
+metadata:
+  name: thing3
+  annotations:
+    optIn: Opting In Again
```
Comment (nicolerenee) [inline test/data/cr_things_filter_update.yml]: and last but not least

File: test/scripts/integration_tests.py [hunk 26]
```diff
@@ -181,34 +255,34 @@ def runTest(self):
...
                 "start",
+                "--nop",
                 "--config",
-                _TEST_DATA_DIRECTORY + "/helm/wait-config.yaml",
-                "--helm-tiller",
-                self.__minikube_ip + ":32664",
-            ],
+                _LOSTROMOS_CONFIGURATION_FILE,
+                "--crd-filter",
+                "optIn",
```
Comment (nicolerenee) [inline test/scripts/integration_tests.py]: Looks like this needs to be changed to match the new filter and then the tests will pass.
